crates/sui-json-rpc/src/object_changes.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::ObjectProvider;
+use sui_json_rpc_types::ObjectChange;
+use sui_types::base_types::{MoveObjectType, SuiAddress};
+use sui_types::governance::StakedSui;
+use sui_types::messages::TransactionEffectsAPI;
+use sui_types::storage::{DeleteKind, WriteKind};
+
+pub async fn get_object_change_from_effect<P: ObjectProvider<Error = E>, E>(
+    object_provider: &P,
+    sender: SuiAddress,
+    effects: &impl TransactionEffectsAPI,
+) -> Result<Vec<ObjectChange>, E> {
+    let mut object_changes = vec![];
+
+    for ((id, version, digest), owner, kind) in effects.all_mutated() {
+        let o = object_provider.get_object(id, version).await?;
+        if let Some(type_) = o.type_() {
+            let type_ = match type_ {
+                MoveObjectType::Other(type_) => Some(type_.clone()),
+                MoveObjectType::StakedSui => Some(StakedSui::type_()),
+                _ => None,
+            };
+
+            if let Some(object_type) = type_ {
+                match kind {
+                    WriteKind::Mutate => object_changes.push(ObjectChange::Mutated {
+                        sender,
+                        owner: *owner,
+                        object_type,
+                        object_id: *id,
+                        version: *version,
+                        digest: *digest,
+                    }),
+                    WriteKind::Create => object_changes.push(ObjectChange::Created {
+                        sender,
+                        owner: *owner,
+                        object_type,
+                        object_id: *id,
+                        version: *version,
+                        digest: *digest,
+                    }),
+                    _ => {}
+                }
+            }
+        } else if let Some(p) = o.data.try_as_package() {
+            if kind == WriteKind::Create {
+                object_changes.push(ObjectChange::Published {
+                    package_id: p.id(),
+                    version: p.version(),
+                    digest: *digest,
+                    modules: p.serialized_module_map().keys().cloned().collect(),
+                })
+            }
+        };
+    }
+
+    for ((id, version, _), kind) in effects.all_deleted() {
+        let o = object_provider
+            .find_object_less_then_version(id, version)
```

Comment:
assuming here it searches previous version, I heard FN will prune previous version aggressively, is it possible that we have a popular object that would update so quickly such that object read here is gone already?

Reply:
yes it could be, in that case it will just return error, just like get_past_object.

Reply:
the transactions table on indexer could be missing some object changes, which is a bit more concerning to me.
last time I heard from phoenix that they will do longer object retentions, I will check with them

Reply:
indexer will not miss it if it is synced within the pruning windows (indexer should know the pruning policy of the full node it is connection to).

the only thing needed to create the object changes are the TransactionEffect and the objects, which we store in the indexer already.

the indexer need to implement the ObjectProvider trait and it will be able to reuse the `get_balance_change_from_effect` methods https://github.com/MystenLabs/sui/blob/2b4523c6a3ca68e9322e6e6319c995c528ea09ef/crates/sui-json-rpc/src/balance_changes.rs#L146-L158